server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParameters.java

```diff
@@ -92,22 +95,53 @@ public enum Metric {
         REPOSITORIES("repositories"),
         ALLOCATIONS("allocations");
 
-        private String metricName;
+        private static final Map<String, Metric> metricMap = Arrays.stream(values())
+            .collect(toUnmodifiableMap(Metric::metricName, Function.identity()));
+        private final String metricName;
+
+        public static final Set<Metric> ALL = Arrays.stream(values()).collect(toUnmodifiableSet());
+        public static final Set<String> ALL_NAMES = metricMap.keySet();
 
         Metric(String name) {
             this.metricName = name;
         }
 
+        public static boolean isValid(String name) {
+            return metricMap.containsKey(name);
+        }
+
+        public static Metric get(String name) {
+            return metricMap.get(name);
+        }
+
+        /**
+         * Read set of metrics from the StreamInput. Ignores unknown metrics.
+         * It should not happen normally, except rolling update when enum changed.
+         * Silently dropping unknown metrics seems less harmful than dropping entire request in this case.
+         */
+        public static Set<Metric> readSetFrom(StreamInput in) throws IOException {
+            final var set = new HashSet<Metric>();
+            final var names = in.readStringArray();
+            for (var name : names) {
+                if (isValid(name)) {
+                    set.add(metricMap.get(name));
```

Comment:
Could we instead switch to sending the proper enum values over the wire, and use this bit only for bwc?

Reply:
I'm not sure how to do that. We do send proper enums within same code version, since we write enums not strings right now. 

For bwc is there trick how to do that? I would assume I need to compare transport version. Do we always increment version when add enum option, how to guarantee that?

Reply:
We can write and read depending on the transport version, something like:
https://github.com/elastic/elasticsearch/blob/3b7b86c507f159180938951e22e611f594325abd/server/src/main/java/org/elasticsearch/cluster/ClusterInfo.java#L116-L120

The version is incremented every time we would like to change a wire format. This is verified by bwc tests or by unit tests if they extend `AbstractWireSerializingTestCase`.

Reply:
>Could we instead switch to sending the proper enum values over the wire, and use this bit only for bwc?

David, could you clarify what do you mean by `proper enum values`? In context of https://github.com/elastic/elasticsearch/pull/107944 I do not think we should rely on `readEnumSet`/`writeEnumSet`.
Did you mean this enum should implement `org.elasticsearch.common.io.stream.Writeable` and we should rely on `org.elasticsearch.common.io.stream.StreamOutput#writeCollection(java.util.Collection<? extends org.elasticsearch.common.io.stream.Writeable>)` ?

Reply:
Yeah I meant `readEnumSet`/`writeEnumSet`. I know they're potentially risky because of BwC but I think we can completely mitigate those risks with a suitable test. I think we should prefer the slimmer wire protocol: even if in this specific case it probably doesn't make much difference, it's the established pattern for sending enum values around.